supervisor/process.go

```diff
@@ -0,0 +1,411 @@
+package supervisor
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pkg/errors"
+	"os/exec"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const (
+	DefaultLineBufferSize = 16384
+	DefaultStreamChanSize = 200
+)
+
+type ProcessState uint8
+
+const (
+	Initial ProcessState = iota
+	Starting
+	Running
+	Stopping
+	Finished
+	Interrupted
+	Fatal
+)
+
+func (s ProcessState) String() string {
+	switch s {
+	case Initial:
+		return "initial"
+	case Starting:
+		return "starting"
+	case Running:
+		return "running"
+	case Stopping:
+		return "stopping"
+	case Finished:
+		return "finished"
+	case Fatal:
+		return "fatal"
+	default:
+		return "unknown"
+	}
+}
+
+// Process is a wrapper around exec.Cmd to run linux processes
+type Process struct {
+	*sync.Mutex
+	cmd          *exec.Cmd
+	streamStdOut chan string
+	streamStdErr chan string
+	stdOut       *StreamOutput
+	stdErr       *StreamOutput
+	state        ProcessState
+	status       Status
+	statusChan   chan *Status
+	done         chan struct{}
+	stateLock    *sync.Mutex
+}
+
+// Status is the status of a process
+type Status struct {
+	Cmd        string
+	PID        int
+	State      ProcessState
+	Error      error
+	ExitCode   int
+	StartTime  int64
+	StopTime   int64
+	StopByUser bool
+}
+
+// NewProcess creates and returns a new process without runnung it
+func NewProcess(name string, args []string) *Process {
+	a := strings.Join(args, " ")
+
+	status := Status{
+		Cmd:        name + " " + a,
+		PID:        0,
+		Error:      nil,
+		ExitCode:   -1,
+		StopByUser: false,
+	}
+
+	cmd := exec.Command(name, args...)
+
+	stdOutChan := make(chan string, DefaultStreamChanSize)
+	stdErrChan := make(chan string, DefaultStreamChanSize)
+	stdOut := NewStreamOutput(stdOutChan)
+	stdErr := NewStreamOutput(stdErrChan)
+
+	process := &Process{
+		cmd:          cmd,
+		status:       status,
+		done:         make(chan struct{}),
+		Mutex:        &sync.Mutex{},
+		stateLock:    &sync.Mutex{},
+		streamStdOut: stdOutChan,
+		streamStdErr: stdErrChan,
+		stdOut:       stdOut,
+		stdErr:       stdErr,
+	}
+
+	return process
+
+}
+
+// setState sets the state of a process
+func (p *Process) setState(state ProcessState) {
+	// cannot the set the state of a finished process
+	if p.state == Finished || p.state == state {
+		return
+	}
+
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	// the next state can only be Starting if the state is Initial
+	if p.state == Initial {
+		p.state = Starting
```

Comment:
This implicit logic may be unexpected for the caller, why not just set the state to whatever is passed in (and have the caller pass proper state)?

Reply:
Yeah I though some basic validation could help the caller avoid mistakes. This logic has been removed